pkg/frontend/admin_openshiftcluster_serialconsole.go

```diff
@@ -0,0 +1,153 @@
+	mgmtstorage "github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/storage"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+)
+
+func (f *frontend) postAdminOpenShiftClusterSerialConsole(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	err := f._postAdminOpenShiftClusterSerialConsole(ctx, w, r, log)
+
+	adminReply(log, w, nil, nil, err)
+}
+
+func (f *frontend) _postAdminOpenShiftClusterSerialConsole(ctx context.Context, w http.ResponseWriter, r *http.Request, log *logrus.Entry) error {
+	vars := mux.Vars(r)
+
+	vmName := r.URL.Query().Get("vmName")
+	err := validateAdminVMName(vmName)
+	if err != nil {
+		return err
+	}
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return err
+	}
+
+	w.Header().Add("Content-Type", "application/octet-stream")
+
+	s, err := newSerialTool(log, f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+
+	return s.dump(ctx, vmName, w)
+}
+
+type serialTool struct {
+	log *logrus.Entry
+	oc  *api.OpenShiftCluster
+
+	accounts        storage.AccountsClient
+	virtualMachines compute.VirtualMachinesClient
+
+	clusterResourceGroup string
+}
+
+func newSerialTool(log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster) (*serialTool, error) {
+	r, err := azure.ParseResourceID(oc.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	fpAuthorizer, err := env.FPAuthorizer(oc.Properties.ServicePrincipalProfile.TenantID, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return &serialTool{
+		log: log,
+		oc:  oc,
+
+		accounts:        storage.NewAccountsClient(r.SubscriptionID, fpAuthorizer),
+		virtualMachines: compute.NewVirtualMachinesClient(r.SubscriptionID, fpAuthorizer),
+
+		clusterResourceGroup: stringutils.LastTokenByte(oc.Properties.ClusterProfile.ResourceGroupID, '/'),
+	}, nil
+}
+
+// dump writes to w the output of the serial console for vmname in raw bytes
+func (s *serialTool) dump(ctx context.Context, vmname string, w io.Writer) error {
```

Comment:
The reason we passed [the `MustGather` GA function](https://github.com/Azure/ARO-RP/blob/43d8b4d0e05b72c218d914e1486afe78999f9531/pkg/frontend/kubeactions/mustgather.go#L24) an `io.Writer` was because the response could be very large and we wanted to avoid a timeout. It's unclear how large the serial console logs could be, but I think the answer might be "very", so I used the same pattern here.